### PR TITLE
Basic http API to fetch different slices of HN items

### PR DESCRIPTION
**Pull request by alexdunne**
This first stage is pretty simple. I've created API using echo, backed by an in-memory store with tests to validate things are being called as expected

### CODE REVIEW

File: cmd/api/main.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	GetAll() hn.Items
+	GetStories() hn.Items
+	GetJobs() hn.Items
+}
+
+func main() {
+	store := inmem.NewInMemoryItemStore()
+
+	server := NewServer(store)
+	server.start("localhost:8000")
```
Comment (jordanfinners) [inline cmd/api/main.go]: I'd make these env vars

File: inmem/InMemoryItemStore.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	return items
+}
+
+type findItemFilter func(item hn.Item) bool
+
+func (s *InMemoryItemStore) find(fn findItemFilter) hn.Items {
+	items := hn.Items{}
+
+	for _, v := range s.store {
+		if fn(v) {
```
Comment (jordanfinners) [inline inmem/InMemoryItemStore.go]: This is a real pet hate of mine, I would prefer just to have the filtering in each of the methods

File: cmd/api/main_test.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+		items = append(items, v)
+	}
+
+	return items
+}
+
+func (s *StubItemStore) GetStories() hn.Items {
+	s.getStoriesCalled++
+
+	items := hn.Items{}
```
Comment (jordanfinners) [inline cmd/api/main_test.go:35]: This is all good, but you might want to look at mocks when you get more services involved, heres an example from realm migration service https://github.com/gymshark/realm-migration-service/blob/40c096f7018e289dd34b74201f9490c859c328f3/internal/mongo/collection_test.go#L21

You can set up the mock, add expected calls and returns and then assert on them

**Review by jordanfinners — CHANGES_REQUESTED** [review]

File: cmd/api/main_test.go (hunk 3 above)
Comment (alexdunne) [reply cmd/api/main_test.go:35]: I looked into using additional testing tools but it seems the vocal go community on the internet are against it :D I'll take a look 👍

File: cmd/api/main.go (hunk 1 above)
Comment (alexdunne) [reply cmd/api/main.go]: 👍 Agreed

**Pull request merged**